src/cpr_data_access/models.py

```diff
@@ -109,11 +111,86 @@ class TextBlock(BaseModel):
     type_confidence: confloat(ge=0, le=1)  # type: ignore
     page_number: conint(ge=-1)  # type: ignore
     coords: Optional[List[Tuple[float, float]]]
+    _spans: list[Span] = PrivateAttr(default_factory=list)
 
     def to_string(self) -> str:
         """Return text in a clean format"""
         return " ".join([line.strip() for line in self.text])
 
+    @property
+    def text_hash(self) -> str:
+        """
+        Get hash of text block text. If the text block has no text (although this shouldn't be the case), return an empty string.
+
+        :return str: md5sum + "__" + sha256, or empty string if the text block has no text
+        """
+
+        if self.text == "":
+            return ""
+
+        text_utf8 = self.to_string().encode("utf-8")
+
+        return (
+            hashlib.md5(text_utf8).hexdigest()
+            + "__"
+            + hashlib.sha256(text_utf8).hexdigest()
+        )
+
+    @property
+    def spans(self) -> Sequence[Span]:
+        """Return all spans in the text block."""
+        return self._spans
+
+    def add_spans(
+        self, spans: Sequence[Span], raise_on_error: bool = False
+    ) -> "TextBlock":
+        """
+        Add spans to the text block.
+
+        If adding spans to a document, `Document.add_spans` should be used instead, as it checks that the document ID of the span matches the text block.
+
+        :param spans: spans to add
+        :param raise_on_error: if True, raise an error if any of the spans do not have `text_block_text_hash` equal to the text block's text hash. If False, print a warning message instead.
+        :raises ValueError: if any of the spans do not have `text_block_text_hash` equal to the text block's text hash
+        :raises ValueError: if the text block has no text
+        :return: text block with spans added
+        """
+
+        LOGGER.warning(
+            "This method should not be used if adding spans to a document as it does not check that the document ID of the span matches document. Use `Document.add_spans` instead."
+        )
```

Comment:
`Span` contains `document_id` which is useful when serialising a `Span` object to disk, and means we can check it's being added to the correct document when calling `Document.add_spans().

However it could also be useful to add a span to a text block, e.g. when experimenting. In this instance the code ignores `document_id` and logs this warning. I'm not sure whether this is the best way to go about things, but I can't think of a way to check `document_id` when using `Document.add_spans()` whilst still enabling `TextBlock.add_spans()`.

Maybe I'm overthinking it and there's no need for `TextBlock.add_spans()`. @stefl14 / @henry-cpr ?

Reply:
this comes down to 'do we care about text blocks not associated with a document'. I'm leaning to no, which means I'd just make `TextBlock.add_spans -> `TextBlock._add_spans`

Reply:
Hmm, agree this is slightly difficult. Seems like an edge case issue that we don't need to think about now for me so would go with latter option